tests/common/gcp_type/firewall_rule_test.py

```diff
@@ -55,6 +55,313 @@ def test_ips_subset_of_ips(self, ips, ips_range, expected):
+                    [{'IPProtocol': 'tcp', 'ports': ['21-23']}]),
+            },
+            firewall_rule.InvalidFirewallRuleError,
+        ),
+        (
+            {
+                'firewall_rule_name': 'n1',
+                'firewall_rule_source_ranges': json.dumps(['1.1.1.1']),
+                'firewall_rule_direction': 'ingress',
+                'firewall_rule_denied': json.dumps(
+                    [{'IPProtocol': 'tcp', 'ports': ['21-23']}]),
+            },
+            firewall_rule.InvalidFirewallRuleError,
+        ),
+        (
+            {
+                'firewall_rule_name': 'n1',
+                'firewall_rule_network': 'n2',
+                'firewall_rule_source_ranges': json.dumps(['1.1.1.1']),
+                'firewall_rule_direction': 'ingress',
+                'firewall_rule_denied': json.dumps(
+                    [
+                        {'IPProtocol': 'tcp', 'ports': ['21-23']},
+                        {},
+                    ]),
+            },
+            firewall_rule.InvalidFirewallActionError,
+        ),
+    ])
+    def test_validate_errors(self, rule_dict, expected_error):
+        rule = firewall_rule.FirewallRule(**rule_dict)
+        with self.assertRaises(expected_error):
+            rule.validate()
+
+    @parameterized.parameterized.expand([
+        (
+            {
+                'firewall_rule_network': 'n2',
+                'firewall_rule_source_ranges': json.dumps(['1.1.1.1']),
+                'firewall_rule_direction': 'ingress',
+                'firewall_rule_network': 'n1',
+                'firewall_rule_denied': json.dumps(
+                    [{'IPProtocol': 'tcp', 'ports': ['21-23']}]),
+            },
+            firewall_rule.InvalidFirewallRuleError,
+        ),
+        (
+            {
+                'firewall_rule_name': 'n1',
+                'firewall_rule_source_ranges': json.dumps(['1.1.1.1']),
+                'firewall_rule_direction': 'ingress',
+                'firewall_rule_denied': json.dumps(
+                    [{'IPProtocol': 'tcp', 'ports': ['21-23']}]),
+            },
+            firewall_rule.InvalidFirewallRuleError,
+        ),
+        (
+            {
+                'firewall_rule_name': 'n1',
+                'firewall_rule_network': 'n2',
+                'firewall_rule_source_ranges': json.dumps(['1.1.1.1']),
+                'firewall_rule_direction': 'ingress',
+                'firewall_rule_denied': json.dumps(
+                    [
+                        {'IPProtocol': 'tcp', 'ports': ['21-23']},
+                        {},
+                    ]),
+            },
+            firewall_rule.InvalidFirewallActionError,
+        ),
+    ])
+    def test_as_json_error(self, rule_dict, expected_error):
+        rule = firewall_rule.FirewallRule(**rule_dict)
+        with self.assertRaises(expected_error):
+            rule.as_json()
+
+    @parameterized.parameterized.expand([
```

Comment:
There is good test data in /tests/inventory/pipelines/test_data/fake_firewall_rules.py. It would be cool to load in EXPECTED_LOADABLE_FIREWALL_RULES and generate the JSON and verify it matches the value in EXPECTED_FIREWALL_RULES_MAP.